### ISSUE TITLE
Add config option to disable public databases when no security config is set

### ISSUE THREAD

**Issue by thigg** [apache/couchdb#3535]
## Summary
Currently the config warns, that if no roles or members are configured, the database will be publicly accessible.

I would propose to add a config switch, which allows to disable this behavior and lets databases without roles or members be fully private.

The motivation behind this could be a scenario like this:
The database holds sensitive data, e.g. from health care. An issue comes up that requires immediate action, e.g. a leak in a client application which allows to get database account data. Now the admin currently on duty at 4am thinks:
"okay, if i just kick all roles with access from the db, I should be fine".

## Desired Behaviour

If the introduced config switch is set and a database has no roles or members configured for access, only admins can access the db.

## Possible Solution
I added a `andalso false` as POC in  https://github.com/apache/couchdb/blob/0059b8f90e58e10b199a4b768a06a762d12a30d3/src/couch/src/couch_db.erl#L700 

This worked in a simple test and denied access, when no roles or members were set, thus the false could be replaced with a config option.

**Comment by rnewson** [comment]
We made this change in https://github.com/apache/couchdb/pull/2339 some time ago.

It is also referenced from the 3.0 release notes https://docs.couchdb.org/en/stable/whatsnew/3.0.html

**Comment by thigg** [comment]
From my experiments and how I understand the PR you added a default _admin role to a newly created DB.
Still, the behavoir is, if you remove all roles and members, the DB is accessible by everyone.

**Comment by rnewson** [comment]
Right, if you _don't_ set the security object yourself, it is implicitly;

```
{
  "members" : { "roles" : [ "_admin" ] },
   "admins" : { "roles" : [ "_admin" ] }
}
```

which means only admins can read/write/etc.

If you _change_ it to `{}`, you are opening it up to everyone.

**Comment by thigg** [comment]
Exactly, this is what the issue is about. 

> The database holds sensitive data, e.g. from health care. An issue comes up that requires immediate action, e.g. a leak in a client application which allows to get database account data. Now the admin currently on duty at 4am thinks:
"okay, if i just kick all roles with access from the db, I should be fine".

CouchDB should help the user to prevent unintentional data leakage and make it hard to release data to the public.

**Comment by rnewson** [comment]
I think what you are asking for is for `{}` to mean private and not public. I sympathise with that position but there is a huge backward compatibility issue there that troubles me. `{}` has had the meaning of "public" in CouchDB since the introduction of the `_users` database (circa 1.2).

The https://github.com/apache/couchdb/pull/2339 PR addressed the surprising "public by default" behaviour in a more careful way, and we phased its rollout over releases.

**Comment by rnewson** [comment]
CouchDB _does_ now help prevent unintential data leakage by making databases private unless changed to public by an administrator.

You are saying that administrators might set databases to public by accident and I agree; we can all make mistakes.

At this point I'm not sure where to go with this issue and invite others to add their thoughts. Would changing the long-standing meaning of the `{}` security object be too disruptive? If we did change it to mean "private", how would one open a database for public read+write?

**Comment by janl** [comment]
I wouldn’t mind adding an off-by-default config option that when enabled means “empty security means admin-only-access”

**Comment by thigg** [comment]
How about adding a config option with a warning, that the default will change in the future.
With some bigger release the default can then be flipped.

**Comment by janl** [comment]
not sure we will address this before we do a (much needed) major revamp, so let’s not worry about future announcements here.

**Comment by janl** [comment]
@thigg are you planning to submit a PR here?

**Comment by thigg** [comment]
No, unfortunately not.
We will just have a proxy in front which checks the authentication separately and blocks unwanted access, thus this risk is mitigated for us anyway.
